WPCOM SSH: Allow user to update an existing SSH key

kozer opened the pull request:
#### Proposed Changes

Right now, when we delete an custom SSH key, the key is detached from all the sites.
So when the user adds a new one, it's needed to go and reassign the ssh key to each site again.

Let the user update the ssh key, solves this problem.

Below are some screenshots of this new functionality.

<img width="1173" alt="image" src="https://user-images.githubusercontent.com/36432/201983246-e9b9c1e2-d206-41e0-a61a-51b220d17644.png">

<img width="1112" alt="image" src="https://user-images.githubusercontent.com/36432/201982901-a25ad65c-80d7-417b-acbf-504004a16380.png">

#### Testing Instructions

##### Test 1

1. Make sure that you apply this patch to your sandbox: D91797-code
2. Sandbox public api.
3. Start calypso dev environment.
4. Navigate to `/me/security/ssh-key`
5. Add an ssh key.
6. Make sure that next to `Remove SSH Key` there is another button, named:`Update SSH Key`
7. Click that button
8. Make sure that a modal displaying `Are you sure you want to update this SSH key? It will affect all attached sites.` appears.
9. Press the OK button.
10. Make sure that a modal displaying the old ssh key and a textarea in order to enter a new one
   appears.
11. Enter a new SSH key and click `Update SSH Key`
12. Ensure that the button changes it's state.
13. After the new key is saved, ensure that the modal is auto closed and a banner indicating that
    the key was saved appears in the top right corner.

###### Test 2

1. Make sure that you apply this patch to your sandbox: D91797-code
2. Sandbox public api.
3. Start calypso dev environment.
4. Navigate to `/me/security/ssh-key`
5. Add an ssh key.
6. Make sure that next to `Remove SSH Key` there is another button, named:`Update SSH Key`
7. Click that button
8. Make sure that a modal displaying `Are you sure you want to update this SSH key? It will affect all attached sites.` appears.
9. Press the OK button.
10. Make sure that a modal displaying the old ssh key and a textarea in order to enter a new one
   appears.
11. Enter a new SSH key and click `Update SSH Key`
12. Ensure that the button changes it's state.
14. Close the modal before the ssh key is saved.
15. Ensure that the `Remove SSH Key button` is disabled.
16. Ensure that the `Update SSH Key` button can be pressed and you can view the modal.
17. Ensure that when the modal reopens, if the update is not yet completed, the textareas and the
    button are in the same state as when the modal was originally closed.

###### Test 3

1. Make sure that you apply this patch to your sandbox: D91797-code
2. Sandbox public api.
3. Start calypso dev environment.
4. Navigate to `/me/security/ssh-key`
5. Add an ssh key.
4. Switch your site to one that has SSH hosting enabled.
5. Navigate to hosting configuration  `/hosting-config/<your domain>` view
6. Make sure to attach the newly added SSH key to the site.
7. Click on your profile icon in the top right corner.
8. Click on `Security -> SSH Key`
9. Update the ssh key as described in `Test 1`
10. Click on `My Sites` on the top left corner
11. Click on `Settings -> Hosting Configuration`
12. Ensure that where previously it was attached the previous SSH key, now there is a loading
    indicator.
13. Wait till the loading is completed.
14. Ensure that the attached SSH key is not the previous one, but the updated SSH key instead


#### Pre-merge Checklist

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

Related to https://github.com/Automattic/dotcom-forge/issues/1135

kozer committed:
Add the ability to change the displaying text in AddSSHKeyForm component

kozer committed:
Add onUpdate property on manage-ssh-keys component

kozer committed:
Add useUpdateSSHKey mutation in order to update an SSH key

kozer committed:
Add the ability to update the ssh key to security-ssh-key component

zaguiini commented:
I haven't tested it in conjunction with the diff. It looks good, and I think that using a modal instead of modifying the existing page makes more sense, because this way the user would be sure that just clicking the "Update SSH key" button won't change anything and they can cancel at any point.

How I imagine this working:

1. User opens the `/me/security/ssh-key` page;
2. User clicks on the "Update SSH key" button;
3. Modal opens with SSH key form, showing which key they are updating;
4. When clicking the submit button inside the modal, the prompt asking "are you sure about that?" pops up;
5. Once clicking yes, the submit button inside the modal shows the loading state;
6. The modal closes and the SSH keys are re-fetched, updating the SSH keys list.

Does that make sense?

Regarding MFA: we're already handling that on previous steps (e.g. when browsing this page/security settings), so I don't think you'll need to implement anything here.

vykes-mac commented:
@zaguiini  @kozer I'm not sure if there was any discussion on this but I'm wondering why we giving users ability to update keys. The examples that I've seen is that a user can only remove an exisitng key and a new one if the want to change it or it expire, Eg. Github, DigitalOcean. I think ssh-keygen doesn't allow update of keys either only adding and removing (_could be wrong_) . Just wondering what benefit we get from updating as oppose to adding a new one which is what most services do, it seems.

kozer commented:
@vykes-mac , yes you are right on that. The only benefit I see here is if we, behind the scenes, also make sure that we reassign the new ssh key to the sites that have assigned the old one. 
If not, it's actually like removing and adding a new one all in one step.
cc: @danielbachhuber

kozer commented:
@zaguiini thanks for your suggestion. The only downside to that is that it will make the mobile experience poorer, but as we have already modals on that page, I think this is ok.

zaguiini commented:
I see @kozer. I think @vykes-mac is right, is not really common to see updates to SSH keys. I think removing and then re-adding, as two different steps for the user, is the way to go. We could even highlight this if necessary.

kozer commented:
@zaguiini So maybe it's better to close the ticket for now? Or leave this open if we need to revisit it?

danielbachhuber commented:
> The examples that I've seen is that a user can only remove an exisitng key and a new one if the want to change it or it expire, Eg. Github, DigitalOcean. I think ssh-keygen doesn't allow update of keys either only adding and removing (_could be wrong_) . Just wondering what benefit we get from updating as oppose to adding a new one which is what most services do, it seems.

@vykes-mac Great observation!

This is something I considered when I created the issue. The key difference between GitHub and WordPress.com is:
1. On GitHub, if you remove your SSH key, adding it again restores SSH access to all of your repositories.
2. On WordPress.com, removing your SSH key also removes it from all attached sites, which means you'll need to reattach it to each site after adding it again.

For this reason, I'd like to continue to implement the original issue.

> How I imagine this working:
> 
> 1. User opens the `/me/security/ssh-key` page;
> 2. User clicks on the "Update SSH key" button;
> 3. Modal opens with SSH key form, showing which key they are updating;
> 4. When clicking the submit button inside the modal, the prompt asking "are you sure about that?" pops up;
> 5. Once clicking yes, the submit button inside the modal shows the loading state;
> 6. The modal closes and the SSH keys are re-fetched, updating the SSH keys list.
> 
> Does that make sense?

@zaguiini I'm fine with this proposed approach. I don't think it's likely for someone to update their SSH key from a mobile device, so fine to switch to a modal.

zaguiini commented:
>2. On WordPress.com, removing your SSH key also removes it from all attached sites, which means you'll need to reattach it to each site after adding it again.
> 
>For this reason, I'd like to continue to implement the original issue.

Just to clarify: the reason why you want to allow updating an SSH key is so one can set a new SSH key without detaching it [the named key] from the existing sites?

danielbachhuber commented:
> the reason why you want to allow updating an SSH key is so one can set a new SSH key without detaching it [the named key] from the existing sites?

@zaguiini Yep!

zaguiini commented:
> > the reason why you want to allow updating an SSH key is so one can set a new SSH key without detaching it [the named key] from the existing sites?
> 
> @zaguiini Yep!

Then one suggestion is to include some copy saying that updating the key means it won't detach from the sites that particular key is being used.

kozer committed:
Change update logic to use a modal

- Show a modal in order to update the ssh key
- Ensure that when the key is updated and navigate back to hosting
  configuration screen the attached ssh keys are reloaded.

kozer commented:
![info-popover-update](https://user-images.githubusercontent.com/497103/201102453-0fa0c24f-243d-4c16-b532-a4d87877f2ef.png)
@danielbachhuber @zaguiini , what do you think on that?

kozer committed:
Add info for the user on why we have the update ssh key functionality

danielbachhuber commented:
@kozer Eh, I don't love that popover. Let's simply incorporate @zaguiini's text suggestion into the existing body of text.

> Also. Do we need to put the freezing strings label here right?

Not in this case. It's not a frequently used UI, so it's fine to display the English strings for a while until the translations are complete.

zaguiini commented:
I'd like for us to display the update copy only when there's an existing key.

Also, the copy looks good:

>Updating an SSH key, it won't detach it from the sites that particular key is being used.

But I think it'd flow better as something like:

>Updating an SSH key won't detach it from the sites that particular key is being used.

I'd also move the copy to the last line, like that:

![image](https://user-images.githubusercontent.com/26530524/201158257-81a2bb58-0f8d-4bf9-a7cf-bb8ad60c2069.png)

I still have mixed feelings about it, though. It looks like a lot of text that I'm not particularly happy about... How about moving the copy to the update form modal?

On another note, I'd appreciate more top spacing here:

![image](https://user-images.githubusercontent.com/26530524/201158593-82925645-096b-49c4-af75-a2916dbdff51.png)

danielbachhuber commented:
> I still have mixed feelings about it, though. It looks like a lot of text that I'm not particularly happy about... How about moving the copy to the update form modal?

💯

kozer commented:
> When the textarea is in an invalid state, the modal stretches out:
> 
> ![image](https://user-images.githubusercontent.com/26530524/201159262-81fce747-1176-43e6-8fd9-b7676fa44612.png)
> 
> One thing I noticed when updating the key is that for a second the modal gets in an intermediary state and bounces before closing. That happens because the existing key gets updated in the component before closing the modal. How about creating a copy of the value through a prop (i.e. `initialValue`) so that when the keys are prefetched, that value doesn't change?

![ssh-modal](https://user-images.githubusercontent.com/497103/201368290-35c4a99e-a6e1-4b1f-95dd-b60496ce0a22.png)
What about that?

I have also some comments/questions:

> When the textarea is in an invalid state, the modal stretches out:
I couldn't find a good way to tackle this. Ohe thing I thought was to setup modal's width to be the same initially as when this text appears. But this makes the modal very wide (almost 1000px) which it wont work on smaller screens. Any ideas?

> One thing I noticed when updating the key is that for a second the modal gets in an intermediary state and bounces before closing. That happens because the existing key gets updated in the component before closing the modal. How about creating a copy of the value through a prop (i.e. initialValue) so that when the keys are prefetched, that value doesn't change?

Can you elaborate a bit more here? It would be really helpful if you show me a video on that.

cc: @zaguiini

kozer committed:
Minor fixes in update-ssh-key-form based on comments.

zaguiini commented:
We're almost there, @kozer!

A couple of things I don't think are ideal:

1. There's no need to display the whole old public key there. How about showing the fingerprint and the key name? The modal gets too tall and even overflows at some point, making the update button disappear from the scrolling area.
2. I feel the update key copy could be a little more prominent, like within a box with a yellow background or something, indicating the information that is pretty important.

>It seems I can't remove the dot. If I do that I get a lint error:  [...] Any way to do so?

Yes, you'll need to pass the translated string as a prop instead of the string you want to translate. So, instead of:

```tsx
updateText="Update SSH Key"
```

and

```tsx
{ sprintf(
  // translators: "sshText" is the text of the ssh save button.
  __( '%(sshText)s.' ),
  {
    sshText: updateText || 'Save SSH Key',
  }
) }
```

It should be:

```tsx
updateText={ __( 'Update SSH Key' ) }
```

and

```tsx
{ updateText ?? __( 'Save SSH Key' ) }
```

In all places (`AddSSHKeyForm` and `UpdateSSHKeyForm`). That will fix it 👍 That error happens because you're trying to translate a string that only contained placeholders, and that doesn't make sense from an i18n perspective.

kozer commented:
> We're almost there, @kozer!
> 
> A couple of things I don't think are ideal:
> 
> 1. There's no need to display the whole old public key there. How about showing the fingerprint and the key name? The modal gets too tall and even overflows at some point, making the update button disappear from the scrolling area.
> 2. I feel the update key copy could be a little more prominent, like within a box with a yellow background or something, indicating the information that is pretty important.
> 
> > It seems I can't remove the dot. If I do that I get a lint error:  [...] Any way to do so?
> 
> Yes, you'll need to pass the translated string as a prop instead of the string you want to translate. So, instead of:
> 
> ```tsx
> updateText="Update SSH Key"
> ```
> 
> and
> 
> ```tsx
> { sprintf(
>   // translators: "sshText" is the text of the ssh save button.
>   __( '%(sshText)s.' ),
>   {
>     sshText: updateText || 'Save SSH Key',
>   }
> ) }
> ```
> 
> It should be:
> 
> ```tsx
> updateText={ __( 'Update SSH Key' ) }
> ```
> 
> and
> 
> ```tsx
> { updateText ?? __( 'Save SSH Key' ) }
> ```
> 
> In all places (`AddSSHKeyForm` and `UpdateSSHKeyForm`). That will fix it +1 That error happens because you're trying to translate a string that only contained placeholders, and that doesn't make sense from an i18n perspective.

What do you think of that @zaguiini ?

![modal-new](https://user-images.githubusercontent.com/497103/201630690-660cb1c4-8422-464e-857e-43245edb1a14.png)

I tried to reuse things we already have for displaying the update key copy

danielbachhuber commented:
@kozer For showing the old SSH key, let's reuse https://github.com/Automattic/wp-calypso/pull/69905

Also, as an alternative to the notice (which looks like an error to me), we could include a title and description for the modal:

> ## Update SSH Key
>
> Update your SSH key to change the SSH key used across all attached sites.

danielbachhuber commented:
> For showing the old SSH key, let's reuse #69905

@kozer This is now available if you want to merge trunk into your feature branch.

kozer committed:
Merge branch 'trunk' into feature/ssh_update_existing_key

kozer commented:
@danielbachhuber 
What do you think of that?
![modal-new](https://user-images.githubusercontent.com/497103/201689861-865a6a92-35bb-4566-b144-cf4f6f6a5fed.png)

danielbachhuber commented:
@kozer Better! I'd like to use a title and description for the modal too, please.

kozer committed:
Minor changes to the update ssh form

- Add description and title to the modal
- Change textarea and use the ssh card component

kozer commented:
@danielbachhuber 
This is the modal updated. Title and description added. However I think that the "warning" alongside with the title are descriptive enough.
If you have any suggestions for the description / title please let me know.
![modal-new-2](https://user-images.githubusercontent.com/497103/201699055-178e4620-a4f8-4d6f-8d6f-327d19382d71.png)

kozer committed:
Updates for update ssh form

- Remove notice
- Add the text of the notice in the description for now

kozer commented:
This is modal for now, after the discussions we had on this thread: p1668440441120489/1668439009.416639-slack-C7YPUHBB2

![modal-new-4](https://user-images.githubusercontent.com/497103/201709613-3395261a-a049-4ff4-be48-40dc7cde34cd.png)

zaguiini commented:
Looking very good and tests nicely 👍 

Non-blocking suggestion: remove the line break between the two sentences inside the modal. Also, is there a way to limit how many lines the text area expands to? Do we need to show the whole key? If we do, how about making it slightly wider so when the key is pasted, less vertical space is consumed?

danielbachhuber commented on client/me/security-ssh-key/security-ssh-key.tsx:
```diff
@@ -44,11 +47,24 @@ const noticeOptions = {
 const sshKeySaveFailureNoticeId = 'ssh-key-save-failure';
+const sshKeyUpdateFailureNoticeId = 'ssh-key-update-failure';
+
+const UpdateSSHModalTitle = styled.h1`
```
Can we use CSS objects, to follow patterns established elsewhere?

danielbachhuber commented on client/me/security-ssh-key/security-ssh-key.tsx:
```diff
@@ -100,6 +116,36 @@ export const SecuritySSHKey = () => {
+		},
+		onSuccess: () => {
+			dispatch( recordTracksEvent( 'calypso_security_ssh_key_update_success' ) );
+			dispatch( successNotice( __( 'SSH key updated to account.' ), noticeOptions ) );
```
```suggestion
			dispatch( successNotice( __( 'SSH key updated for account.' ), noticeOptions ) );
```

danielbachhuber commented on client/me/security-ssh-key/security-ssh-key.tsx:
```diff
@@ -100,6 +116,36 @@ export const SecuritySSHKey = () => {
+			dispatch( removeNotice( sshKeyUpdateFailureNoticeId ) );
+		},
+		onSuccess: () => {
+			dispatch( recordTracksEvent( 'calypso_security_ssh_key_update_success' ) );
```
Can you create pull requests to record these events with Tracks?

kozer replied:
Here is the pr for that

Update success:
https://github.com/Automattic/tracks-events-registration/pull/1175

Update failure
https://github.com/Automattic/tracks-events-registration/pull/1176

danielbachhuber commented on client/me/security-ssh-key/manage-ssh-keys.tsx:
```diff
@@ -25,6 +32,17 @@ const SSHKey = ( { userLogin, sshKey, onDelete, keyBeingDeleted }: SSHKeyProps )
 	};
 
+	const handleUpdateClick = () => {
+		accept(
```
Instead of having two modals, which is a little bit of an odd experience, let's simply add a 'Cancel' button to the modal with the form:

<img width="804" alt="image" src="https://user-images.githubusercontent.com/36432/201770789-a4fb7090-1116-48bc-8133-cddc5d2e1fce.png">

One modal would be much better.

danielbachhuber commented on client/me/security-ssh-key/manage-ssh-keys.tsx:
```diff
@@ -46,8 +64,16 @@ const SSHKey = ( { userLogin, sshKey, onDelete, keyBeingDeleted }: SSHKeyProps )
 				</SSHKeyCard.Date>
 			</SSHKeyCard.Details>
 			<SSHKeyCard.Button
+				primary={ true }
```
This button shows up as red for me...

<img width="1093" alt="image" src="https://user-images.githubusercontent.com/36432/201771001-55a7d940-140a-4eab-b3dd-372ef9ff337d.png">

We should probably make 'Remove SSH key' the primary action, and 'Update SSH key' the secondary.

danielbachhuber commented on client/me/security-ssh-key/security-ssh-key.tsx:
```diff
@@ -155,6 +201,35 @@ export const SecuritySSHKey = () => {
+					>
+						<div style={ { width: '600px' } }>
+							<UpdateSSHModalTitle>{ __( 'Update SSH Key' ) }</UpdateSSHModalTitle>
+							<UpdateSSHModalDescription>
```
For the SSH modal description, let's use:
> Replace your current SSH key with a new SSH key to use the new SSH key with all attached sites.

danielbachhuber commented on client/me/security-ssh-key/update-ssh-key-form.tsx:
```diff
@@ -0,0 +1,126 @@
+			} }
+		>
+			<FormFieldset>
+				<FormLabel>{ __( 'Old SSH Public Key' ) }</FormLabel>
```
```suggestion
				<FormLabel>{ __( 'Current SSH Public Key' ) }</FormLabel>
```

danielbachhuber commented on client/me/security-ssh-key/update-ssh-key-form.tsx:
```diff
@@ -0,0 +1,126 @@
+					</SSHKeyCard.Details>
+				</SSHKeyCard.Root>
+				<NewSSHFormFieldContainer>
+					<FormLabel htmlFor={ PUBLIC_SSH_KEY_INPUT_ID }>{ __( 'SSH Public Key' ) }</FormLabel>
```
```suggestion
					<FormLabel htmlFor={ PUBLIC_SSH_KEY_INPUT_ID }>{ __( 'New SSH Public Key' ) }</FormLabel>
```

danielbachhuber commented on client/me/security-ssh-key/update-ssh-key-form.tsx:
```diff
@@ -0,0 +1,126 @@
+	keyName: string;
+}
+
+export const UpdateSSHKeyForm = ( {
```
It seems like there's a lot of duplication between this and `<AddSSHKeyForm>`.

Is there any way we could consolidate both to a more generic `<SSHKeyForm>`?

kozer replied:
They have many things in common, however, they have different structures. I believe in keeping it KISS instead of trying a more general component with strange props to keep it DRY.
If we found that there is a need to reuse the same logic to other places as well, I agree that maybe it worth the effort to put more thought into that.

danielbachhuber reviewed (commented):
Solid progress, @kozer ! Left a few comments to work through.

kozer committed:
Fix update ssh button style

kozer committed:
Update wording

kozer committed:
Update styled components to object

kozer committed:
Add cancel button to the update ssh form

kozer committed:
Merge branch 'trunk' into feature/ssh_update_existing_key

kozer committed:
Update ssh update's form modal width

danielbachhuber committed:
Use named endpoint for PUT request

danielbachhuber committed:
Disable the 'Update SSH key' button if the key is being deleted

danielbachhuber committed:
Use logical properties for RTL support

sejas commented on client/me/security-ssh-key/manage-ssh-keys.tsx:
```diff
@@ -46,8 +57,18 @@ const SSHKey = ( { userLogin, sshKey, onDelete, keyBeingDeleted }: SSHKeyProps )
+				primary={ true }
+				scary={ false }
+				disabled={ !! keyBeingDeleted || !! keyBeingUpdated }
+				onClick={ handleUpdateClick }
+				style={ { marginInlineStart: 'auto' } }
+			>
+				{ __( 'Update SSH key' ) }
+			</SSHKeyCard.Button>
+			<SSHKeyCard.Button
+				style={ { marginInlineStart: '10px' } }
 				busy={ keyBeingDeleted === sshKey.name }
-				disabled={ !! keyBeingDeleted }
+				disabled={ !! keyBeingDeleted || !! keyBeingUpdated }
```
Since `keyBeingUpdated` is used only to calculate if the buttons should be disabled, maybe we could pass disabled (`!! keyBeingDeleted || !! keyBeingUpdated`) as a prop to `SSHKey` component, and the same variable can be used in both buttons. Line 62 and Line 71.

sejas reviewed (approved):
@kozer  Congrats !! It's looking great !!
I've added a tiny optimization totally optional.

Tested with the API D91797-code and both are working nicely ! Well done !! 👏

danielbachhuber committed:
Disable the 'Cancel' button when the SSH key is being updated

danielbachhuber reviewed (approved):
Nicely done on this, @kozer ! I really like the end result.

I made a few changes directly to your PR:
* [Use named endpoint for PUT request](https://github.com/Automattic/wp-calypso/pull/69826/commits/355aa2bdf0f99ffb5fc1baf0022624dc164306f8)
* [Disable the 'Update SSH key' button if the key is being deleted](https://github.com/Automattic/wp-calypso/pull/69826/commits/7ee118a6d139bacf647a3acbc61f58b1bcb9235d)
* [Use logical properties for RTL support](https://github.com/Automattic/wp-calypso/pull/69826/commits/27d3edc9b306f7cd1814f6770f4660dc8384f396)
* [Disable the 'Cancel' button when the SSH key is being updated](https://github.com/Automattic/wp-calypso/pull/69826/commits/6fbb99fcfda382b41f7a32786e3cf5960a972a07)

I think this is ready to ship after you land D91797-code

kozer commented:
> Nicely done on this, @kozer ! I really like the end result.
> 
> I made a few changes directly to your PR:
> 
> * [Use named endpoint for PUT request](https://github.com/Automattic/wp-calypso/pull/69826/commits/355aa2bdf0f99ffb5fc1baf0022624dc164306f8)
> * [Disable the 'Update SSH key' button if the key is being deleted](https://github.com/Automattic/wp-calypso/pull/69826/commits/7ee118a6d139bacf647a3acbc61f58b1bcb9235d)
> * [Use logical properties for RTL support](https://github.com/Automattic/wp-calypso/pull/69826/commits/27d3edc9b306f7cd1814f6770f4660dc8384f396)
> * [Disable the 'Cancel' button when the SSH key is being updated](https://github.com/Automattic/wp-calypso/pull/69826/commits/6fbb99fcfda382b41f7a32786e3cf5960a972a07)
> 
> I think this is ready to ship after you land D91797-code

Thanks @danielbachhuber for your effort!

kozer committed:
Refactor disable buttons expression to a var in manage ssh keys
component

kozer committed:
Merge branch 'trunk' into feature/ssh_update_existing_key

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7625922

Thank you @kozer for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7625922).
